File: phone_combos.py
def is_valid(d, s):
    keypad = [set(),
              set(),
              set(["A","B","C"]),
              set(["D","E","F"]),
              set(["G","H","I"]),
              set(["J","K","L"]),
              set(["M","N","O"]),
              set(["P","Q","R","S"]),
              set(["T","U","V"]),
              set(["W","X","Y","Z"])]
    lhs = 0
    rhs = 0
    while lhs < len(d) and not len(keypad[int(d[lhs])]):
        lhs += 1
    while lhs < len(d):
        digit = int(d[lhs])
        if len(keypad[digit]):
            if s[rhs] in keypad[digit]:
                rhs += 1
            else:
                return False
        lhs += 1
    return lhs == len(d) and rhs == len(s)


def helper(keypad, digits, index, curr, results):
    if index == len(digits):
        results.append(''.join(curr))
        return
    digit = int(digits[index])
    if not len(keypad[digit]):
        helper(keypad, digits, index+1, curr, results)
    else:
        for letter in keypad[digit]:
            curr.append(letter)
            helper(keypad, digits, index+1, curr, results)
            curr.pop()
    return


def all_perm(d):
    results = list()
    curr = list()
    keypad = [set(),
              set(),
              set(["A","B","C"]),
              set(["D","E","F"]),
              set(["G","H","I"]),
              set(["J","K","L"]),
              set(["M","N","O"]),
              set(["P","Q","R","S"]),
              set(["T","U","V"]),
              set(["W","X","Y","Z"])]
    helper(keypad, d, 0, curr, results)
    return results


def main():
    assert is_valid("000000001","") == True
    assert is_valid("0000000012","A") == True
    assert is_valid("2000000001","A") == True
    assert is_valid("234","ADG") == True
    assert is_valid("","AASDKWE") == False
    assert is_valid("012345678900","ADGJMPTW") == True
    assert is_valid("012345678900","DAGJMPTW") == False
    print(all_perm("23"))


if __name__ == "__main__":
    main()

File: phone_combos.cpp
/*
Problem: Given two strings, one containing only digits, the second containing
         only letters, determine if the letter string is a valid interpretation
         of the digits string

Example:
         Input:  "234", "ADG"
         Output: True


--------------------------------------------------------------------------------

Problem: Given a string containing only digits, return all possible string
         combinations that are valid interpretations of the digits string

Example:
         Input:  "2"
         Output: ["a", "b", "c"]

KEYPAD:
2   ABC
3   DEF
4   GHI
5   JKL
6   MNO
7   PQRS
8   TUV
9   WXYZ

*/

#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

bool valid_combo(std::string n, std::string s)
{
    if (n.empty() != s.empty()) return false;
    static const std::unordered_map<char,std::string> keypad {{'0', ""},
                                                              {'1', ""},
                                                              {'2', "ABC"},
                                                              {'3', "DEF"},
                                                              {'4', "GHI"},
                                                              {'5', "JKL"},
                                                              {'6', "MNO"},
                                                              {'7', "PQRS"},
                                                              {'8', "TUV"},
                                                              {'9', "WXYZ"}};
    size_t lhs = 0;
    size_t rhs = 0;
    while (lhs < n.size() && rhs < s.size()) {
        std::string temp = keypad.at(n[lhs]);
        if (temp.size()
            && std::find_if(temp.begin(),
                            temp.end(),
                            [&s, &rhs](char c){ return c == s[rhs]; }) != temp.end()) {
            ++lhs;
            ++rhs;
        }
        else if (!temp.size()) {
            ++lhs;
        }
    }
    return lhs == n.size() && rhs == s.size();
}


int main()
{
    assert(valid_combo("234", "ADG"));
    assert(valid_combo("1111111112", "A"));
    assert(!valid_combo("1111111112", "AA"));
    assert(valid_combo("00000001", ""));
    return 0;
}
